geonode/maps/api/views.py

```diff
@@ -86,12 +92,122 @@ def local_datasets(self, request, pk=None):
 
     def perform_create(self, serializer):
-        if serializer.is_valid():
-            hookset().perform_create(self, serializer)
-            return serializer
+        self._get_data_from_blob(serializer)
+        # Thumbnail will be handled later
+        post_creation_data = {"thumbnail": serializer.validated_data.pop("thumbnail_url", "")}
+        # M2M maplayers
+        serializer.validated_data["maplayers"] = self._create_m2m_maplayers(serializer)
+
+        instance = serializer.save(
+            owner=self.request.user,
+            resource_type="map",
+            uuid=str(uuid4()),
+        )
+
+        # thumbnail, events and resouce routines
+        self._post_change_routines(
+            instance=instance,
+            create_action_perfomed=True,
+            additional_data=post_creation_data,
+        )
 
     def perform_update(self, serializer):
-        """Associate current user as task owner"""
-        if serializer.is_valid():
-            hookset().perform_update(self, serializer)
-            return serializer
+        # Check instance permissions with resolve_object
+        mapid = serializer.validated_data["id"]
+        key = "urlsuffix" if Map.objects.filter(urlsuffix=mapid).exists() else "pk"
+        map_obj = resolve_object(self.request, Map, {key: mapid}, permission="base.change_resourcebase", permission_msg=_PERMISSION_MSG_SAVE)
+        instance = serializer.instance
+        assert instance == map_obj
+        # Thumbnail will be handled later
+        post_change_data = {
+            "thumbnail": serializer.validated_data.pop("thumbnail_url", ""),
+            "dataset_names_before_changes": [lyr.alternate for lyr in instance.local_datasets],
+        }
+        # M2M maplayers
+        serializer.validated_data["maplayers"] = self._create_m2m_maplayers(serializer)
+
+        instance = serializer.save()
+
+        # thumbnail, events and resouce routines
+        self._post_change_routines(
+            instance=instance,
+            create_action_perfomed=False,
+            additional_data=post_change_data,
+        )
+
+    def _create_m2m_maplayers(self, serializer):
+        if "maplayers" not in serializer.validated_data:
+            # Do nothing, partial update, without changes to maplayers
+            return []
+
+        if serializer.instance:
+            # Delete all existing maplayers
+            serializer.instance.maplayers.all().delete()
+
+        # Initialize maplayers serializer
+        maplayers_data = serializer.validated_data.pop("maplayers")
+        maplayers_serializer = MapLayerSerializer(data=maplayers_data, many=True)
+        maplayers_serializer.is_valid(raise_exception=True)
+
+        # Create new maplayers, map relation will be added by serializer.save()
+        maplayers = maplayers_serializer.save()
+        return maplayers
+
+    def _get_maplayers_data_from_blob(self, blob_map_data: dict):
```

Comment:
@italogsfernandes why are you extracting this information from the blob? We said we don't want GeoNode to put its nose into the blob. And, again, I don't see why we need to extract this data. Where are they required?

Reply:
@giohappy The blob operations are inside `_get_data_from_blob` and `_get_maplayers_data_from_blob`.
The current version of Mapstore does not send the the `maplayers` data outside the blob. Without this information the relation between Maps and Layers won't be created at geonode.

This needs to change when mapstore will be up to date and that part is here, for the moment, exists only to ensure that the maplayers relation will be created when the `maplayers` field is not provided bu the POST/PATCH.

Reply:
@italogsfernandes I'd prefer to remove this from the beginning. GeoNode currently doesn't rely on this information, since we're not showing the map / maplayer / dataset relationships now. If we keep them we risk to hide potential issues.

Reply:
So, it's removed. I started the `test_suite` to see if it will broke something not seem by us yet.